bundler/helpers/lib/functions/lockfile_updater.rb

```diff
@@ -0,0 +1,261 @@
+      end
+
+      specs.include?(spec_name)
+    end
+
+    # Copied from Bundler::Runtime: Modified to only prune gems that have
+    # been unlocked
+    def prune_gem_cache(resolve, cache_path, unlocked_gems)
+      cached_gems = Dir["#{cache_path}/*.gem"]
+
+      outdated_gems = cached_gems.reject do |path|
+        spec = Bundler.rubygems.spec_from_gem path
+
+        !unlocked_gems.include?(spec.name) || resolve.any? do |s|
+          s.name == spec.name && s.version == spec.version &&
+            !s.source.is_a?(Bundler::Source::Git)
+        end
+      end
+
+      return unless outdated_gems.any?
+
+      outdated_gems.each do |path|
+        File.delete(path)
+      end
+    end
+
+    # Copied from Bundler::Runtime
+    def prune_git_and_path_cache(resolve, cache_path)
+      cached_git_and_path = Dir["#{cache_path}/*/.bundlecache"]
+
+      outdated_git_and_path = cached_git_and_path.reject do |path|
+        name = File.basename(File.dirname(path))
+
+        resolve.any? do |s|
+          s.source.respond_to?(:app_cache_dirname) &&
+            s.source.app_cache_dirname == name
+        end
+      end
+
+      return unless outdated_git_and_path.any?
+
+      outdated_git_and_path.each do |path|
+        path = File.dirname(path)
+        FileUtils.rm_rf(path)
+      end
+    end
+
+    def unlock_yanked_gem(dependencies_to_unlock, error)
+      raise unless error.message.match?(GEM_NOT_FOUND_ERROR_REGEX)
+
+      gem_name = error.message.match(GEM_NOT_FOUND_ERROR_REGEX).
+                named_captures["name"]
+      raise if dependencies_to_unlock.include?(gem_name)
+
+      dependencies_to_unlock << gem_name
+    end
+
+    # rubocop:disable Metrics/PerceivedComplexity
+    def unlock_blocking_subdeps(dependencies_to_unlock, error)
+      all_deps =  Bundler::LockfileParser.new(lockfile).
+                  specs.map(&:name).map(&:to_s)
+      top_level = build_definition([]).dependencies.
+                  map(&:name).map(&:to_s)
+      allowed_new_unlocks = all_deps - top_level - dependencies_to_unlock
+
+      raise if allowed_new_unlocks.none?
+
+      # Unlock any sub-dependencies that Bundler reports caused the
+      # conflict
+      potentials_deps =
+        error.cause.conflicts.values.
+        flat_map(&:requirement_trees).
+        map do |tree|
+          tree.find { |req| allowed_new_unlocks.include?(req.name) }
+        end.compact.map(&:name)
+
+      # If there are specific dependencies we can unlock, unlock them
+      if potentials_deps.any?
+        return dependencies_to_unlock.append(*potentials_deps)
+      end
+
+      # Fall back to unlocking *all* sub-dependencies. This is required
+      # because Bundler's VersionConflict objects don't include enough
+      # information to chart the full path through all conflicts unwound
+      dependencies_to_unlock.append(*allowed_new_unlocks)
+    end
+    # rubocop:enable Metrics/PerceivedComplexity
+
+    def build_definition(dependencies_to_unlock)
```

Comment:
Several of the function classes implement this, there's some light variance in arguments but the gist of the implementation looks very close.

This is maybe a candidate to extract and unit test?

Reply:
We should punt on this for now as it means further transforms on a large/tricky extraction